File: Olimpiady/OIJ:OIG/||| etap/Centrala_telefoniczna.cpp
#include <bits/stdc++.h>

using namespace std;

long double odleglosc(int y,int x){
    return sqrt(pow(y,2)+pow(x,2));
}
long long koszt(int h){
    return (h-1)*h/2;
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    
    priority_queue < pair <double, long long> > kolejka;
    long long ilosc_domow,zwiekszanie,ilosc_wierzcholkow,zaplata,x,y,wynik=0,zysk=0,wysokosc;
    long double maksi;
    pair <long double,long long> punkt;
    cin >> ilosc_domow >> zwiekszanie;
    for (int i=0; i<ilosc_domow; i++){
        maksi = (double)0;
        cin >> ilosc_wierzcholkow >> zaplata;
        for (int i=0; i<ilosc_wierzcholkow*2; i+=2){
            cin >> x >> y;
            maksi = max(maksi,(long double)odleglosc(y,x));
        }
        kolejka.push({-maksi,zaplata});
    }
    
    for (int i=0; i<ilosc_domow; i++){
        punkt = kolejka.top();
        kolejka.pop();
        zysk += punkt.second;
        wysokosc = ceil((punkt.first*-1)/zwiekszanie);
        wynik = max(wynik,zysk-koszt(wysokosc));
    }
    cout << wynik;
    return 0;
}

File: Olimpiady/OIJ:OIG/||| etap/Centrala_telefoniczna.py
import queue,math
kolejka  = queue.PriorityQueue()
ilosc_domow,zwiekszanie = map(int,input().split())
def odleglosc(y,x):
    return math.sqrt(math.pow(y,2)+math.pow(x,2))

def koszt(h):
    return (h-1)*h//2

for i in range(ilosc_domow):
    dane = input().split()
    maxi = 0
    for k in range(2,int(dane[0])*2+1,2):
        maxi = max(maxi,odleglosc(int(dane[k+1]),int(dane[k])))
    kolejka.put((maxi,int(dane[1])))
wynik = 0
zysk = 0
for i in range(kolejka.qsize()):
    punkt = kolejka.get()
    zysk += punkt[1]
    print(wynik,punkt[0],punkt[1])
    wysokosc = math.ceil(punkt[0]/zwiekszanie)
    wynik = max(wynik,zysk-koszt(wysokosc))

print(wynik)

